### PR TITLE
chore(coderd/notifications): expand golden file testing for notifications

### LINKED ISSUE #14913: Notifications: verify rendered action
The directory [rendered-templates](https://github.com/coder/coder/tree/e6d8f674ad48a601f25bb98420a14f1150914ead/coderd/notifications/testdata/rendered-templates) contains bodies and titles. We could extend it to also cover actions (in JSON format?).

### PR DESCRIPTION
**Pull request by SasSwart**
This PR aims to close https://github.com/coder/coder/issues/14913.

It expands the golden files for the notifier to include the entire payload serialised as JSON.

### CODE REVIEW

File: coderd/notifications/notifications_test.go [hunk 1]
```diff
@@ -700,7 +699,9 @@ func TestNotificationTemplates_Golden(t *testing.T) {
 			name: "TemplateWorkspaceDeleted",
 			id:   notifications.TemplateWorkspaceDeleted,
 			payload: types.MessagePayload{
-				UserName: "Bobby",
+				UserName:     "Bobby",
```
Comment (SasSwart) [inline coderd/notifications/notifications_test.go:702]: Note to self: Should I still be using a payload struct in these test cases? We do use all their fields in the test, but we don't use the payload struct itself.

File: coderd/notifications/notifications_test.go [hunk 2]
```diff
@@ -762,7 +771,9 @@ func TestNotificationTemplates_Golden(t *testing.T) {
 			name: "TemplateUserAccountCreated",
 			id:   notifications.TemplateUserAccountCreated,
 			payload: types.MessagePayload{
-				UserName: "Bobby",
+				UserName:     "Bobby",
+				UserEmail:    "bobby@coder.com",
+				UserUsername: "bobby",
 				Labels: map[string]string{
 					"created_account_name":      "bobby",
 					"created_account_user_name": "William Tables",
```
Comment (SasSwart) [inline coderd/notifications/notifications_test.go:779]: Note to self: Given `UserName` here and `created_account_user_name` a few lines below. Do I need both?

File: coderd/notifications/notifications_test.go [hunk 3]
```diff
@@ -1285,6 +1330,49 @@ func (f *fakeHandler) Dispatcher(payload types.MessagePayload, _, _ string) (dis
...
+
+func (f *goldenFileHandler) Dispatcher(payload types.MessagePayload, title, body string) (dispatch.DeliveryFunc, error) {
+	return func(_ context.Context, _ uuid.UUID) (retryable bool, err error) {
+		partialName := strings.Split(f.t.Name(), "/")[1]
+		goldenFile := filepath.Join("testdata", "rendered-templates", partialName+".json.golden")
+
+		// UserIDs change on every test run. We need to set it to a known value to compare golden files.
+		payload.UserID = "00000000-0000-0000-0000-000000000000"
+
+		// The following is done by the stmp handler during dispatch.
```
Comment (SasSwart) [inline coderd/notifications/notifications_test.go:1346]: Note to self: improve wording here

File: coderd/notifications/notifications_test.go [hunk 4]
```diff
@@ -1285,6 +1330,49 @@ func (f *fakeHandler) Dispatcher(payload types.MessagePayload, _, _ string) (dis
...
+		// what would have been sent. We don't need to maintain parity
+		// with how the SMTP handler does it. We just need an accurate
+		// representation of the payload, and the rendered body and title.
+		payload.Labels["_body"] = body
+		payload.Labels["_title"] = title
+
+		payloadJSON, err := json.MarshalIndent(payload, "", "  ")
+		require.NoError(f.t, err, "want no error marshaling payload to JSON")
+
+		if *updateGoldenFiles {
```
Comment (SasSwart) [inline coderd/notifications/notifications_test.go:1357]: Note to self: This should either refer to the struct attribute or we should remove the struct attribute

File: coderd/notifications/testdata/rendered-templates/TemplateTemplateDeleted.json.golden [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (SasSwart) [inline coderd/notifications/testdata/rendered-templates/TemplateTemplateDeleted.json.golden:1]: Note to self: This serialisation of the payload with rendered body and title is useful. It proves that all the right things are passed to the dispatcher. However, it doesn't prove that the smtp dispatcher renders the actions correctly. I should consider adding golden tests to the smtp dispatcher

File: coderd/notifications/dispatch/smtp_test.go [hunk 6]
```diff
@@ -523,18 +525,129 @@ func TestSMTP(t *testing.T) {
 	}
 }
 
-func pingClient(listen net.Listener, useTLS bool, startTLS bool) (*smtp.Client, error) {
-	tlsCfg := &tls.Config{
-		// nolint:gosec // It's a test.
-		InsecureSkipVerify: true,
+func TestSMTPGolden(t *testing.T) {
```
Comment (SasSwart) [inline coderd/notifications/dispatch/smtp_test.go]: Note to self: remove this test. it's defunct

File: coderd/notifications/dispatch/testdata/rendered-templates/TemplateWorkspaceDeleted.golden.html [hunk 7]
```diff
@@ -0,0 +1,66 @@
+From: system@coder.com
+To: bobby@coder.com
+Subject: Workspace bobby-workspace has been deleted
+Message-Id: 74472365-4868-463c-8325-1a0c99d12ac0@blush-whale-48
```
Comment (SasSwart) [inline coderd/notifications/dispatch/testdata/rendered-templates/TemplateWorkspaceDeleted.golden.html:4]: note to self: set all uuids to uuid.Nil

File: coderd/notifications/dispatch/testdata/rendered-templates/TemplateWorkspaceDeleted.golden.html [hunk 8]
```diff
@@ -0,0 +1,66 @@
+From: system@coder.com
+To: bobby@coder.com
+Subject: Workspace bobby-workspace has been deleted
+Message-Id: 74472365-4868-463c-8325-1a0c99d12ac0@blush-whale-48
+Date: Thu, 10 Oct 2024 10:42:15 +0000
+Content-Type: multipart/alternative;  boundary=32122c97f1062aa05b08887844a8756eb74e6b2de42c5bdcab2584aed1fb
```
Comment (SasSwart) [inline coderd/notifications/dispatch/testdata/rendered-templates/TemplateWorkspaceDeleted.golden.html:6]: note to self: filter out the message boundary

File: coderd/notifications/dispatch/testdata/rendered-templates/TemplateWorkspaceDeleted.golden.html [hunk 9]
```diff
@@ -0,0 +1,66 @@
+From: system@coder.com
+To: bobby@coder.com
+Subject: Workspace bobby-workspace has been deleted
+Message-Id: 74472365-4868-463c-8325-1a0c99d12ac0@blush-whale-48
+Date: Thu, 10 Oct 2024 10:42:15 +0000
```
Comment (SasSwart) [inline coderd/notifications/dispatch/testdata/rendered-templates/TemplateWorkspaceDeleted.golden.html:5]: note to self: replace the date with a constant

File: coderd/notifications/notifications_test.go [hunk 10]
```diff
@@ -938,53 +985,263 @@ func TestNotificationTemplates_Golden(t *testing.T) {
...
+					smtpCfg,
+					*db,
+					defaultHelpers(),
+					createMetrics(),
+					logger.Named("manager"),
+				)
+				require.NoError(t, err)
+
+				smtpManager.Run(ctx)
+				time.Sleep(time.Second)
```
Comment (mafredri) [inline coderd/notifications/notifications_test.go:1086]: Can we avoid this sleep somehow?

File: coderd/notifications/notifications_test.go [hunk 11]
```diff
@@ -938,53 +985,263 @@ func TestNotificationTemplates_Golden(t *testing.T) {
...
+					err = os.MkdirAll(filepath.Dir(goldenFile), 0o755)
+					require.NoError(t, err, "want no error creating golden file directory")
+					err = os.WriteFile(goldenFile, body, 0o600)
+					require.NoError(t, err, "want no error writing body golden file")
+					return
+				}
+
+				wantBody, err := os.ReadFile(goldenFile)
+				require.NoError(t, err, fmt.Sprintf("missing golden notification body file. %s", hint))
+				require.Equal(t, string(wantBody), string(body), fmt.Sprintf("smtp notification does not match golden file. If this is expected, %s", hint))
```
Comment (mafredri) [inline coderd/notifications/notifications_test.go:1135]: Suggestion: In my experience require has pretty hard to parse diffs, `cmp.Diff` can produce a bit nicer output (it's used in a few other places too).

File: coderd/notifications/notifications_test.go [hunk 12]
```diff
@@ -938,53 +985,263 @@ func TestNotificationTemplates_Golden(t *testing.T) {
...
+				t.Parallel()
+				// nolint:gocritic // Unit test.
+				ctx := dbauthz.AsSystemRestricted(testutil.Context(t, testutil.WaitSuperLong))
+
+				// Spin up the mock webhook server
+				server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					w.WriteHeader(http.StatusOK)
+
+					body, err := io.ReadAll(r.Body)
+					require.NoError(t, err)
```
Comment (mafredri) [inline coderd/notifications/notifications_test.go:1148]: Nit: Since the handler runs in a separate goroutine, we can't use `require` here as it calls `t.FailNow` (https://pkg.go.dev/testing#F.FailNow).

You could either move handling back to test routine via channel or use assert+return instead. We should also make sure that the handler is done before we exit the test, especially if we keep logic in it (i.e. verify if Close waits for all handlers to complete or immediately stops the server and ignores the handlers continued execution).

**Review by mafredri — COMMENTED** [review]
Nice work on increasing the test coverage. I didn't look too closely at the actual rendered output as I lack a bit of context, but perhaps @dannykopping or @mtojek has more thoughts on that.

File: coderd/notifications/testdata/rendered-templates/smtp/TemplateTemplateDeleted.html.golden [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+Content-Transfer-Encoding: quoted-printable
+Content-Type: text/plain; charset=UTF-8
+
+Hi Bobby,
+
+The template bobby-template was deleted by rob.
+
+The template's display name was Bobby's Template.
+
+
```
[the remark this answers is not in the document]
Comment (SasSwart) [reply coderd/notifications/testdata/rendered-templates/smtp/TemplateTemplateDeleted.html.golden:19]: I'm not sure. This is how it renders. I haven't changed anything about the whitespace consciously. As far as I can tell this must be how it has always rendered. Should I reduce it?

File: coderd/database/migrations/000263_consistent_notification_initiator_naming.down.sql [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+WHERE
+	id = '4e19c0ac-94e1-4532-9515-d1801aa283b2';
+
+-- UserAccountDeleted
+UPDATE notification_templates
+SET
+	body_template = E'Hi {{.UserName}},\n\n' ||
+					E'User account **{{.Labels.deleted_account_name}}** has been deleted.\n\n' ||
+					-- Mention the real name of the user who deleted the account:
+					E'The deleted account belonged to **{{.Labels.deleted_account_user_name}}** and was deleted by **{{.Labels.account_deleter_user_name}}**.'
```
[the remark this answers is not in the document]
Comment (mafredri) [reply coderd/database/migrations/000263_consistent_notification_initiator_naming.down.sql:17]: Oh right, I missed that 😄, very nice!

File: coderd/notifications/testdata/rendered-templates/smtp/TemplateTemplateDeleted.html.golden (hunk 13 above)
[the remark this answers is not in the document]
Comment (mafredri) [reply coderd/notifications/testdata/rendered-templates/smtp/TemplateTemplateDeleted.html.golden:19]: If that's the case, no need to change it on my account, let's see what @mtojek thinks.

File: coderd/notifications/notifications_test.go (hunk 12 above)
Comment (SasSwart) [reply coderd/notifications/notifications_test.go]: Fixed

File: coderd/notifications/notifications_test.go (hunk 11 above)
Comment (SasSwart) [reply coderd/notifications/notifications_test.go]: done

File: coderd/notifications/notifications_test.go (hunk 10 above)
Comment (SasSwart) [reply coderd/notifications/notifications_test.go]: fixed

File: coderd/notifications/dispatch/mocksmtp/server.go [hunk 15]
```diff
@@ -1,4 +1,4 @@
-package dispatch_test
```
[the remark this answers is not in the document]
Comment (SasSwart) [reply coderd/notifications/dispatch/mocksmtp/server.go]: done

File: coderd/notifications/notifications_test.go [hunk 16]
```diff
@@ -938,53 +986,306 @@ func TestNotificationTemplates_Golden(t *testing.T) {
...
+
+				// Wait for the message to be fetched
+				var msg *smtptest.Message
+				require.Eventually(t, func() bool {
+					msg = backend.LastMessage()
+					return msg != nil && len(msg.Contents) > 0
+				}, testutil.WaitShort, testutil.IntervalFast)
+
+				// require.NotNil(t, msg, "want a message to be sent")
+				// require.NotEmpty(t, msg.Contents, "want a non-empty message")
```
Comment (mtojek) [inline coderd/notifications/notifications_test.go:1120]: nit: dead code?

File: coderd/notifications/notifications_test.go [hunk 17]
```diff
@@ -938,53 +986,306 @@ func TestNotificationTemplates_Golden(t *testing.T) {
...
+
+	content = dateRegex.ReplaceAll(content, []byte("Date: "+constantDate))
+	content = messageIDRegex.ReplaceAll(content, []byte("Message-Id: "+constantMessageID))
+	content = bytes.ReplaceAll(content, boundary, []byte(constantBoundary))
+
+	return content
+}
+
+func normalizeGoldenWebhook(content []byte) []byte {
+	const constantUUID = "00000000-0000-0000-0000-000000000000"
```
Comment (mtojek) [inline coderd/notifications/notifications_test.go:1282]:
```
{
  "_version": "1.1",
  "msg_id": "00000000-0000-0000-0000-000000000000",
  "payload": {
    "_version": "1.1",
    "notification_name": "Workspace Autobuild Failed",
    "notification_template_id": "00000000-0000-0000-0000-000000000000",
    "user_id": "00000000-0000-0000-0000-000000000000",
```

Ok, I know why you did this but I'm wondering if there is a way to keep stable uuids written to the golden file 🤔. It could help find out why some labels change, and prevent bugs in the future.

PS. If this is difficult->impossible, it can be postponed for later (add FIXME label).

File: coderd/notifications/notifications_test.go [hunk 18]
```diff
@@ -25,8 +27,13 @@ import (
 
 	"golang.org/x/xerrors"
 
+	"cdr.dev/slog"
+
+	"github.com/emersion/go-sasl"
+
```
Comment (mtojek) [inline coderd/notifications/notifications_test.go:33]: nit: there are some empty lines between imports

File: coderd/notifications/testdata/rendered-templates/smtp/TemplateTemplateDeleted.html.golden (hunk 13 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/notifications/testdata/rendered-templates/smtp/TemplateTemplateDeleted.html.golden:19]: Most likely it can be removed, but let's keep it aside from this PR.

File: coderd/notifications/testdata/rendered-templates/smtp/TemplateWorkspaceMarkedForDeletion.html.golden [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+Message-Id: 02ee4935-73be-4fa1-a290-ff9999026b13@blush-whale-48
+Date: Fri, 11 Oct 2024 09:03:06 +0000
+Content-Type: multipart/alternative;  boundary=bbe61b741255b6098bb6b3c1f41b885773df633cb18d2a3002b68e4bc9c4
+MIME-Version: 1.0
+
+--bbe61b741255b6098bb6b3c1f41b885773df633cb18d2a3002b68e4bc9c4
+Content-Transfer-Encoding: quoted-printable
+Content-Type: text/plain; charset=UTF-8
+
+Hi Bobby,
```
Comment (mtojek) [inline coderd/notifications/testdata/rendered-templates/smtp/TemplateWorkspaceMarkedForDeletion.html.golden:13]: I didn't know that we could render two content-types. Curious, if we can store golden results in separate files?

**Review by mtojek — APPROVED** [review]
I believe this is in a good shape 👍 Left a couple of nit-picks to take a look & reply. Feel free to ignore them or postpone for later!

Nice job, Sas 👍 👍

File: coderd/notifications/testdata/rendered-templates/smtp/TemplateWorkspaceMarkedForDeletion.html.golden (hunk 19 above)
Comment (SasSwart) [reply coderd/notifications/testdata/rendered-templates/smtp/TemplateWorkspaceMarkedForDeletion.html.golden:13]: > Curious, if we can store golden results in separate files?

The premise that I adopted for this PR was one of integration testing. This output is meant to represent exactly what would have been sent to the customer. I did not make them render in the same file. This is how it has been.

To separate them would, in my opinion, introduce artificial opportunity for drift from actual behaviour. These tests are most reliable if they not only mimick, but actually dogfood the actual system's behaviour.

If you'd like me to change it to be multiple files, I would be happy to. Would you mind helping me understand the benefit though?

File: coderd/notifications/notifications_test.go (hunk 18 above)
Comment (SasSwart) [reply coderd/notifications/notifications_test.go]: fixed.

File: coderd/notifications/notifications_test.go (hunk 17 above)
Comment (mtojek) [reply coderd/notifications/notifications_test.go:1274]: I mean, we could replace dynamic UUIDs with some const/known values rather than zeros. For instance:

user_id -> `ceb6824e-7fe1-4000-aad9-b80bb141560a`

File: coderd/notifications/testdata/rendered-templates/smtp/TemplateWorkspaceMarkedForDeletion.html.golden (hunk 19 above)
Comment (mtojek) [reply coderd/notifications/testdata/rendered-templates/smtp/TemplateWorkspaceMarkedForDeletion.html.golden:13]: Ok! If there are no benefits, then let's leave them as is. Maybe replace `.html.golden` with `.smtp.golden` as this is not just HTML file, but also Plaintext and headers.

> Would you mind helping me understand the benefit though?
My original intention was keep HTML in `.html.golden` files, and Plaintext in `.txt.golden`

File: coderd/notifications/testdata/rendered-templates/webhook/TemplateUserAccountActivated.json.golden [hunk 20]
```diff
@@ -0,0 +1,29 @@
+{
+  "_version": "1.1",
+  "msg_id": "00000000-0000-0000-0000-000000000000",
+  "payload": {
+    "_version": "1.1",
+    "notification_name": "User account activated",
+    "notification_template_id": "00000000-0000-0000-0000-000000000000",
```
Comment (mtojek) [inline coderd/notifications/testdata/rendered-templates/webhook/TemplateUserAccountActivated.json.golden:7]: One more comment:

Can we keep `notification_template_id` as is, and not zero it? These values are const/stable.

**Review by mtojek — APPROVED** [review]
Let's get this large PR merged, nit-picks can be done in a follow-up.

File: coderd/notifications/notifications_test.go (hunk 17 above)
Comment (SasSwart) [reply coderd/notifications/notifications_test.go:1274]: ACK. Thanks
